tests/test_redis_brokers.py

```diff
@@ -78,22 +110,138 @@ def test_running_job(broker):
     assert broker.get_jobs_from_queue('foo_queue', 1) == []
 
 
+def test_cant_exceed_max_concurrency(broker):
+    set_concurrency_keys(broker)
+    # An idempotent job is required.
+    job1 = Job(
+        CONCURRENT_TASK_NAME, 'foo_queue', datetime.now(timezone.utc), 1,
+        # kwargs help with debugging but are not part of the test.
+        task_kwargs=dict(name='job1'),
+    )
+    job2 = Job(
+        CONCURRENT_TASK_NAME, 'foo_queue', datetime.now(timezone.utc), 1,
+        task_kwargs=dict(name='job2'),
+    )
+    broker.enqueue_jobs([job1, job2])
+    returned_jobs = broker.get_jobs_from_queue('foo_queue', 2)
+    assert len(returned_jobs) == CONCURRENT_TASK_MAX_CONCURRENCY
+    assert returned_jobs[0].task_kwargs == dict(name='job1')
+
+    # Check the current concurrency was set properly.
+    current = broker._r.hget(
+        broker._to_namespaced(CURRENT_CONCURRENCY_KEY), CONCURRENT_TASK_NAME
+    )
+    assert current == b'1'
+
+    # Make sure job2 was left alone in the queue.
+    queued = broker._r.lpop(broker._to_namespaced('foo_queue'))
+    assert json.loads(queued)['id'] == str(job2.id)
+
+
+def test_get_jobs_from_queue_returns_all_requested(broker):
+    # If a job is not returned because it was over concurrency limits,
+    # make sure the number of jobs requested is filled from other jobs
+    # and that the over-limit one is left alone in the queue.
+    set_concurrency_keys(broker)
+    jobs = [
+        Job(CONCURRENT_TASK_NAME, 'foo_queue', datetime.now(timezone.utc), 1),
+        Job(CONCURRENT_TASK_NAME, 'foo_queue', datetime.now(timezone.utc), 1),
+        Job('foo_task', 'foo_queue', datetime.now(timezone.utc), 1),
+    ]
+    broker.enqueue_jobs(jobs)
+    returned_jobs = broker.get_jobs_from_queue('foo_queue', 2)
+    assert len(returned_jobs) == CONCURRENT_TASK_MAX_CONCURRENCY + 1
+
+
+def test_set_concurrency_keys_sets_new_keys_on_idempotent_tasks(broker):
+    max_c = random.randint(1, 10)
```

Comment:
Could you set it to a fixed int? In my experience randomness is best avoided in tests.

Reply:
It's not the first time I've had this from reviewers - an Openstack guy said not to use `random`, but was perfectly happy with a random uuid instead... go figure!

My own experience is the exact opposite, it is a great way to show test _intent_ - that is we take any int and it works. I have a ~38k LOC Python project that uses Spinach based on a test factory that generates random test inputs where required for 1761 tests. Occasionally we find bugs due to the randomness as well.

I will change it if you insist, but honestly it seems quite benign here and seems a nice way to show that the test will work with low integers. Let me know if I convinced you or not and I will change if not.

Reply:
I'm fine with leaving it as is but I'm not convinced. I'm not denying that introducing randomness helps uncovering bugs, but in my mind there is a clear distinction between a test suite and fuzzing.

I find immense value in knowing that one version of the code always leads to the same result in CI, a bit like seeing the test suite as a pure function: the same code as input always gives the same output.

Reply:
I just had a similar conversation on Twitter with someone I've known for a long time. He said it was fuzzing too, but it's not fuzzing. In this particular case, it is demonstrating that an input value makes it to an output. With a fixed input, you are not testing properly because the code could actually have a bug that results in the same fixed output and unless you vary the input you don't catch it. I have actually seen this happen!

There's two main reasons for using randomness. The reason I just gave above, and also where you need _a_ value for something because it's a required input, but it's not part of the test, so you just need _something_ but don't care what it is. That is an important thing to show in tests as _test intent_.

Anyway I think I have dragged this out too far now!